README.md
# ml2vim
Fork of ml2vim (found [here](https://library.wolfram.com/infocenter/MathSource/2584/) ) by Timo Felbinger.

I fixed some compatibility issues with modern MathLink installation and deprecated function calls.

Description taken from the original site:

This is a Mathematica front end built into Vim, which is a high-end programmer's editor, highly compatible to vi but with lots of additional features. The front end is just an add-on for the editor, which can still be used as a standard editor for all kinds of ASCII files.

The front end is a "text-mode' application which will run in any terminal (including, of course, xterms). The functionality is similar to the standard notebook interface, without graphics (there is, however, built-in support for external viewers like ghostview), but with greatly enhanced editing capabilities. A sample file can be found at http://www.quantum.physik.uni-potsdam.de/TF/samplevma.html The editor has extensive online documentation (covering the front end, too).

You need two packages: The editor Vim with the front end extensions (current version: vim-5.4g-mma0.10), and a small auxiliary program, ml2vim, which translates between Vim and the kernel (current version: ml2vim-0.99.8).

Both packages come as source code (precompiled packages for Linux/Intel and True64 Unix/Alpha are available, too) and can be copied freely. Vim is charityware (see README.txt for details), ml2vim is covered by the GNU General Public license.

Only UNIX(-like) systems are supported. So far, this front end has been successfully tested on Linux/Intel and True64 Unix/Alpha.


HISTORY

19990202
  First error-free build of a "proof-of-concept": 0.99.0

19990218
  First alpha version released: 0.99.7

  Refined the protocol used when talking to Vim:
  Meta-messages REFRESH and EOC are defined and will be handled.

  Added options: 
    --piggyback, --refresh

  Added support for command line options:
    --help, --version, --debug
  A trace() mechanism is defined which will produce lots of
  diagnostic output when --debug is given on the command line.

  Many minor bugs eliminated.

  Added support for viewers

  Added option --suppress-... ,  removed obsolete option --display

  Added include-mechanism (similar to C's #include)
  This required another refinement of the protocol: Tags may now,
  in certain cases, include ranges of numbers instead of single
  numbers.

19990429
  New release: 0.99.8

19990500
  Update for compatibility with newer MathLink and replacement of deprecated function calls.



utc.c
/*  utc.c
 *
 *  Read and set system time in canonical format.
 *
 *  Copyright (C) 1998 Timo Felbinger 
 *  email: Timo.Felbinger@quantum.physik.uni-potsdam.de
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program (in a file called LICENSE); if not, write to 
 *  the Free Software Foundation, Inc., 59 Temple Place - Suite 330, 
 *  Boston, MA 02111-1307, USA, or look at http://www.gnu.org.
 */

#define PROGRAM "utc"

int debug = 0;

#include "jupp.c"

int main( int argc, char *(argv[]) ) {
  struct tm newtime;
  /* time_t newcaltime; */
  struct timespec ts;
  static char utc_buffer[16];

  if( argc <= 1 ) {
    utc_now( utc_buffer );
    fputs( utc_buffer, stdout );
    fflush( stdout );
    return 0;
  }

  if( strlen( argv[1] ) != 15 ) {
    ufprintf( stderr, "utc: argument not in canonical format.\n" );
    fflush( stderr );
    return 1;
  }
  if( argv[1][8] != '.' ) {
    if( argv[1][8] != ':' ) {
      ufprintf( stderr, "utc: argument not in canonical format.\n" );
      fflush( stderr );
      return 1;
    }
  }

  putenv( "TZ=:UTC" );

  /*   The following call of tzset is redundant under Linux, but
   *   required eg by Digital Unix:
   */
  tzset();

  sscanf( argv[1] + 13, "%u", &(newtime.tm_sec) );
  argv[1][13] = 0;
  sscanf( argv[1] + 11, "%u", &(newtime.tm_min) );
  argv[1][11] = 0;
  sscanf( argv[1] +  9, "%u", &(newtime.tm_hour) );
  argv[1][ 9] = 0;
  sscanf( argv[1] +  6, "%u", &(newtime.tm_mday) );
  argv[1][ 6] = 0;
  sscanf( argv[1] +  4, "%u", &(newtime.tm_mon) );
  newtime.tm_mon--;
  argv[1][ 4] = 0;
  sscanf( argv[1], "%u", &(newtime.tm_year) );
  newtime.tm_year -= 1900;
  newtime.tm_isdst = 0;

  errno = 0;
  ts.tv_nsec = 0;
  /* newcaltime = mktime( &newtime ); */
  /* if( newcaltime != -1 ) { */
  if(1) {
    errno = 0;
    if( clock_settime(CLOCK_REALTIME, &ts) ) {
      ufprintf( stderr, "utc: clock_settime() failed: %s", strerror(errno) );
      fflush( stderr );
      return -1;
    } else {
      return 0;
    }
  } else {
    ufprintf( stderr, "utc: mktime() failed: argument format screwed up?" );
    fflush( stderr );
    return -1;
  }

  return 0;
}



include/double2decimal.c
/* double2decimal.c    
 *
 *  Copyright (C) 1998 Timo Felbinger 
 *  email: Timo.Felbinger@quantum.physik.uni-potsdam.de
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program (in a file called LICENSE); if not, write to 
 *  the Free Software Foundation, Inc., 59 Temple Place - Suite 330, 
 *  Boston, MA 02111-1307, USA, or look at http://www.gnu.org.
 *
 *
 *   double2decimal: converts a double number into decimal notation
 *
 *   the result is returned as a mantissa and decimal exponent.
 *   for regular (neither NAN nor INF) numbers, the mantissa
 *   will be a NUL-terminated string of exactly FP_MAX_PRECISION
 *   digits. the first digit is not a '0' unless the converted
 *   value itself was 0.0. the decimal point of the mantissa is
 *   to the left of the leftmost digit. no sign is included in the
 *   string; rather, a separate flag will indicate negative numbers.
 *
 *   the conversion will always be performed up to the maximum number
 *   of significant digits in a double. the task of rounding, if
 *   desired, is left to the caller.
 *
 *   for non-regular numbers, the string will contain either "INF"
 *   or "NAN";  +INF and -INF are indicated by the sign flag.
 *   NANs have no meaningful sign.
 */

#ifndef __DOUBLE2DECIMAL_C
#define __DOUBLE2DECIMAL_C

#include "math.h"

#define _DEXPLEN        11     //  same as djgppv2
#define _HIDDENBIT      1      //  same as djgppv2
#define _FEXPLEN        8      //  same as djgppv2 
#define BITS(type)      (BITSPERBYTE * (int)sizeof(type)) 
#define DSIGNIF (BITS(double) - _DEXPLEN + _HIDDENBIT - 1) 

/*  maximum number of significant digits (15 for IEEE 64-bit numbers): */
#define FP_MAX_PRECISION  ( (DSIGNIF * 3) / 10 )

/*  we demand the caller to provide a buffer for the result,
 *  to make this function thread-safe.
 */
struct double2decimal_buffer {
  char dummy[ FP_MAX_PRECISION + 1 ];
};

void double2decimal(
  double value    /*  value to be converted  */
, struct double2decimal_buffer *the_buffer  /*  buffer for the created string (mantissa) */
, int *pdp        /*  position of the decimal point (i.e., the decimal exponent) */
                  /*  (0 means: left of the leftmost digit in the buffer) */
, int *pneg       /*  flag: number is negative (maybe -INF) */
, int *pregular   /*  flag: number is neither NAN nor INF */
) {
  int i;
  double ipart;
  char *const buf = (char *)the_buffer;
  char *p, *pe;

  static int initialized = 0;
  static double bigval, inv_bigval;
  if( ! initialized ) {
    double v = 10.0;  /* for thread safety, we use a auto variable for intermediate result */
    for( i = 1; i < FP_MAX_PRECISION; i++ ) {
      v *= 10.0;
    }
    bigval = v;
    inv_bigval = 1.0 / v;
    initialized = 1;
  }

  if( isnan( value ) ) {
    *pregular = 0;
    strcpy( buf, "NAN" );
    *pneg = 0;    /* not really meaningful, but we avoid printing -NAN */
    return;
  }
  if( ( i = isinf( value ) ) ) {
    *pregular = 0;
    *pneg = ( i < 0 );  /*  -INF and +INF are really different */
    strcpy( buf, "INF" );
    return;
  }
  *pregular = 1;

  if( value < 0 ) 
    *pneg = 1, value = -value;
  else
    *pneg = 0;

  /*  we get the number into the range [ inv_bigval .. 1 [ and remember
   *  the decimal exponent, by multiplying with very large or small factors:
   */
  *pdp = 0;
  p = buf;
  pe = buf + FP_MAX_PRECISION;
  if( value > 0 ) {
    while( value < inv_bigval ) {
      value *= bigval;
      *pdp -= FP_MAX_PRECISION;
    }
    while( value >= 1 ) {
      value *= inv_bigval;
      *pdp += FP_MAX_PRECISION;
    }
  } else {    /*  shortcut for value == 0.0  */
    for( ; p < pe ; *(p++) = '0' )
      ;
    *pe = NUL;
    return;
  }

  /*  we now have a number in [ inv_bigval .. 1 [ ; we normalize it
   *  so it is in the range [ 1 .. 10 [:
   */
  for( ;; ) {
    value = modf( value * 10, &ipart );
    i = (int) ipart;
    if( i ) 
      break;
    else
      -- *pdp;
  }

  /*  we now have the correct decimal exponent and can compute the 
   *  mantissa (we compute one additional digit for rounding):
   */
  while( p <= pe ) {
    *(p++) = i + '0';
    value = modf( value * 10, &ipart );
    i = (int) ipart;
  }

  /*  do the rounding:  */
  p = pe;
  *pe += 5;
  while( *p > '9' ) {
    *p = '0';
    if( p > buf )
      ++ * --p; 
    else                   /*  we have rounded up /all/ digits, so we now have only '0's; so... */
      *p = '1', (*pdp)++;  /*  ...we make it a `1000000....'  */
  }

  /*  the last digit (which had been used only for rounding) becomes the terminating NUL: */
  *pe = NUL;
}

#endif



include/jupp.c
/* jupp.c
 *
 *  Copyright (C) 1998 Timo Felbinger 
 *  email: Timo.Felbinger@quantum.physik.uni-potsdam.de
 *
 *  This file contains standard macros, classes and functions, and
 *  it handles a lot of early stuff. It is the first file to be
 *  #included in most of my programs. this is a stripped version
 *  containing stuff required by ml2vim.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program (in a file called LICENSE); if not, write to 
 *  the Free Software Foundation, Inc., 59 Temple Place - Suite 330, 
 *  Boston, MA 02111-1307, USA, or look at http://www.gnu.org.
 */


#ifndef __JUPP_C
#define __JUPP_C

/*
 *  config.h should have been created by configure from config.h.in:
 */
#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include <fcntl.h>
#include <time.h>
#include <sys/time.h>
#include <ctype.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <errno.h>
#include <values.h>
#include <getopt.h>
#include <signal.h>

#define NO    0
#define FALSE 0
#define YES   1
#define TRUE  1
#define NUL   '\0'
#define TheCowsComeHome NO

#define until(X)  while(!(X))

#if ! HAVE_ISINF
#  define isinf(X) ( !isnan(X) && !finite(X) )
#endif

/*
 *  low-level error handling
 */

/*  counter for general error conditions: */
int problems = 0;

/*  counter for signals caught and handled: */
int signals_handled = 0;

/*  this value should be only increased via |= and finally be returned by main: */
int retval = 0;
/*  use the following constants to |= into retval: */
#define RETVAL_UNDEFINED  0x0001
#define RETVAL_MEMORY     0x0002
#define RETVAL_FP         0x0004
#define RETVAL_SYSCALL    0x0008
#define RETVAL_LOWLEVEL   0x0010
#define RETVAL_SEGFAULT   0x0020
#define RETVAL_USER       0x0040
#define RETVAL_SECURITY   0x0080

/*
 *  a more flexible replacement for the ?printf family of functions
 *  (this also solves the problem of missing family members on some
 *   systems; eg, Digital Unix 4.0c doesn't know snprintf): 
 */
#include "double2decimal.c"
#include "uprintf.c"

/*  create date-time-string in canonical 15 characters + '\0' format
 *  (buffer s must take 16 bytes). this may create a Y10K problem...
 */
void utc_date_time( char *s, time_t ctm ) {
  struct tm *ptm;

  ptm = gmtime( &ctm );
  usnprintf( 
    s, 16, "%04u%02u%02u.%02u%02u%02u"
    , 1900 + ptm->tm_year, (ptm->tm_mon)+1, ptm->tm_mday
    , ptm->tm_hour, ptm->tm_min, ptm->tm_sec
  );
}
char *utc_now( char *s ) {
  utc_date_time( s, time(NULL) );
  return s;
}

FILE *stdlog = NULL;
FILE *childlog = NULL;
void trace ( char *fmt, ... ) {
  if( debug ) {
    char now[16];
    va_list vap;
    va_start( vap, fmt );
    utc_now( now );
    if( ! stdlog )
      stdlog = fopen( "ml2vim.log", "a" );
    if( stdlog ) {
      ufprintf( stdlog, "\n%s: ", now );
      vufprintf( stdlog, fmt, vap );
      if( errno )
        ufprintf( stdlog, "  errno:%d", errno );
      fflush( stdlog );
    }
    va_end( vap );
  }
}
void ctrace ( char *fmt, ... ) {
  if( debug ) {
    char now[16];
    va_list vap;
    va_start( vap, fmt );
    utc_now( now );
    if( ! childlog )
      childlog = fopen( "ml2vim_child.log", "a" );
    if( childlog ) {
      ufprintf( childlog, "\n%s: ", now );
      vufprintf( childlog, fmt, vap );
      if( errno )
        ufprintf( childlog, "  errno:%d", errno );
      fflush( childlog );
    }
    va_end( vap );
  }
}

/*  a function to deal with serious errors at an early stage where 
 *  more advanced exception handling may not yet be available (if it
 *  is implemented at all).
 */
void die( const char * msg, int err ) {
  if( debug ) {
    trace( "%s: %\\s: %d\n", PROGRAM, "", msg, err );
    trace( "%s: exiting", PROGRAM );
    fclose( stdlog );
  }
  ufprintf( stderr, "\n%s: %\\s: %d\n", PROGRAM, "", msg, err );
  exit( retval );
}

void * safe_malloc( size_t n ) {
  void *p = malloc( n );
  if( ! p )
    /*  we seem to be short on memory, so we probably can't do much  */
    /*  about it, so we just exit:                                   */
    die( "malloc() failed", (int)n );
  return p;
}

void safe_free( void ** pp ) {
  free( *pp );
  *pp = NULL;
}

/*
 *  exec_child: replace current process by the given cmdline.
 *  This function is guaranteed to never return.
 *  The command line will be parsed into individual arguments;
 *  arguments are
 *    - either any sequence of nonblank characters, separated by whitespace,
 *    - or quoted arguments, starting and ending with a `"'; a backslash 
 *      escapes any following char except NUL, or starts an octal escape 
 *      sequence of up to 3 octal digits.
 */
void exec_child( char * cmdline ) {
  char **argv = NULL;
  int argc;
  int pass;
  int quoted;
  int escaped;
  unsigned char c;
  size_t len;
  char *p;

  /*  pass 1:  count arguments, mallocate argv                    */
  /*  pass 2:  count argument lengths, mallocate argv's elements  */
  /*  pass 3:  copy arguments                                     */

  ctrace( "exec_child: got cmdline >>>%s<<<", cmdline );

  for( pass = 1; pass <= 3; ++pass ) {

    ctrace( "exec_child: starting pass %d", pass );

    for( p = cmdline, argc = 0 ; *p ; ++argc ) {
      if( ( quoted = ( *p == '"' ) ) ) ++p;

      len = 0;
      do {
        if( ( escaped = ( *p == '\\' ) ) ) ++p;
        if( ! *p ) break;

        if( escaped ) {

          if( isdigit( *p ) ) {
            c = (unsigned char)*(p++) - (unsigned char)'0';
            if( isdigit( *p ) ) {
              c = c * 8 + (unsigned char)*(p++) - (unsigned char)'0';
              if( isdigit( *p ) )
                c = c * 8 + (unsigned char)*(p++) - (unsigned char)'0';
            }
            if( pass == 3 ) argv[ argc ] [ len ] = (char)c;
          } else {
            if( pass == 3 ) argv[ argc ] [ len ] = *p;
            ++p;
          }
          ++len;

        } else {

          if( quoted ? (*p == '"') : isspace( *p ) ) break;
          if( pass == 3 ) argv[ argc ] [ len ] = *p;
          ++len, ++p;

        }

      } while( TRUE );  /*  loop over one argument  */
      if( pass == 3 )
        argv[ argc ] [ len ] = NUL;

      if( pass == 2 ) 
        argv[ argc ] = (char *)safe_malloc( len + 1 );
      while( *p && isspace( *p ) ) 
        ++p;
    }  /*  loop over arguments  */

    if( pass == 1 )
      argv = (char **)safe_malloc( (size_t)( (argc + 1) * sizeof(char *) ) );
  }  /*  loop over passes  */
  argv[ argc ] = NULL;

  ctrace( "exec_child: about to execute:" );
  for( argc = 0; argv[argc]; ++argc ) {
    ctrace( "  >>>%s<<<", argv[argc] );
  }
  execvp( *argv, argv );
  exit( 2 );
}

#endif

